cmd/snap-update-ns/secure.go

```diff
@@ -0,0 +1,187 @@
+func (as *Assumptions) MockUnrestrictedPrefixes(prefixes ...string) (restore func()) {
+	old := as.unrestrictedPrefixes
+	as.unrestrictedPrefixes = prefixes
+	return func() {
+		as.unrestrictedPrefixes = old
+	}
+}
+
+// AddChange records the fact that a change was applied to the system.
+func (as *Assumptions) AddChange(change *Change) {
+	as.pastChanges = append(as.pastChanges, change)
+}
+
+// isRestricted returns true if a path follows restricted writing scheme.
+//
+// Writing to a restricted path results in step-by-step validation of each
+// directory, starting from the root of the file system. Unless writing is
+// allowed a mimic must be constructed to ensure that writes are not visible in
+// undesired locations of the host filesystem.
+//
+// Provided path is the full, absolute path of the entity that needs to be
+// created (directory, file or symbolic link).
+func (as *Assumptions) isRestricted(path string) bool {
+	// Anything rooted at one of the unrestricted prefixes is not restricted.
+	// Those are for things like /var/snap/, for example.
+	for _, prefix := range as.unrestrictedPrefixes {
+		if strings.HasPrefix(path, prefix) {
+			return false
+		}
+	}
+	// All other paths are restricted
+	return true
+}
+
+// canWriteToDirectory returns true if writing to a given directory is allowed.
+//
+// Writing is allowed in one of thee cases:
+// 1) The directory is in one of the explicitly permitted locations.
+//    This is the strongest permission as it explicitly allows writing to
+//    places that may show up on the host, one of the examples being $SNAP_DATA.
+// 2) The directory is on a read-only filesystem.
+// 3) The directory is on a tmpfs created by snapd.
+func (as *Assumptions) canWriteToDirectory(dirFd int, dirName string) (bool, error) {
+	if !as.isRestricted(dirName) {
+		return true, nil
+	}
+	var fsData syscall.Statfs_t
+	if err := sysFstatfs(dirFd, &fsData); err != nil {
+		return false, fmt.Errorf("cannot fstatfs %q: %s", dirName, err)
+	}
+	// Writing to read only directories is allowed because EROFS is handled
+	// by each of the writing helpers already.
+	if ok := IsReadOnly(dirFd, dirName, &fsData); ok {
+		return true, nil
+	}
+	// Writing to a trusted tmpfs is allowed because those are not leaking to
+	// the host.
+	if ok := IsSnapdCreatedPrivateTmpfs(dirFd, dirName, &fsData, as.pastChanges); ok {
+		return true, nil
+	}
+	// If writing is not not allowed by one of the three rules above then it is
+	// disallowed.
+	return false, nil
+}
+
+// RestrictionsFor computes restrictions for the desired path.
+func (as *Assumptions) RestrictionsFor(desiredPath string) *Restrictions {
+	if as.isRestricted(desiredPath) {
+		return &Restrictions{assumptions: as, desiredPath: desiredPath, restricted: true}
+	}
+	return nil
+}
+
+// TrespassingError is an error when filesystem operation would affect the host.
+type TrespassingError struct {
+	ViolatedPath string
```

Comment:
Maybe we can drop `Path` suffix as it's kind assumed already. BTW. `Violated` looks a bit awkward, wondering if there could be a nicer alternative.

Reply:
How about `Intended` and `Blocked`?